dsref/parse.go

```diff
@@ -0,0 +1,162 @@
+package dsref
+
+import (
+	"fmt"
+	"regexp"
+)
+
+// These functions parse a string to create a dsref. We refer to a "human-friendly reference"
+// as one with only a username and dataset name, such as "my_user/my_dataset". A "full reference"
+// can also contain a "concrete reference", which includes an optional profileID plus a network
+// and a "commit hash".
+//
+// Parse will parse a human-friendly reference, or only a concrete reference, or a full reference.
+//
+// ParseHumanFriendly will only successfully parse a human-friendly reference, and nothing else.
+//
+// The grammar is here:
+//
+//  <dsref> = <humanFriendlyRef> [ <concreteRef> ] | <concreteRef>
+//  <humanFriendlyRef> = <username> '/' <datasetname>
+//  <concreteRef> = '@' [ <profileID> ] '/' <network> '/' <commitHash>
```

Comment:
This is great. I think we should merge as-is because it's a better, tested implementation of the way we currently think about ref parsing. 

I'd like to change a few things about the grammar for reference strings in the future, and would like to structure these as follow-up PRs, using this one as a baseline

* we need consistent naming for the components of a reference string. in some places we call the left side an `Alias`, others the `Name`, and here we're now adding `human-friendly reference`. 
* We need to introduce grammar for reference strings that use logbook identifiers.
* profileID should move within the grammar, and be a variant of a human friendly name. 

Lots to think about, just wanted to surface here so we can revisit this PR later.